File: 498-diagonal-traverse/diagonal-traverse.ts
function findDiagonalOrder(mat: number[][]): number[] {
    if (mat.length === 0) return []
    const m: number = mat.length
    const n: number = mat[0].length
    const result: number[] = []
    const diagonals: number[][] = Array.from({ length: m + n - 1 }, () => [])

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            const key: number = i + j
            diagonals[key].push(mat[i][j])
        }
    }

    for (let k = 0; k < diagonals.length; k++) {
        const diagonal: number[] = diagonals[k]
        if (k % 2 === 0) {
            for (let p = diagonal.length - 1; p >= 0; p--) {
                result.push(diagonal[p])
            }
        } else {
            for (let p = 0; p < diagonal.length; p++) {
                result.push(diagonal[p])
            }
        }
    }

    return result
};

File: 498-diagonal-traverse/diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        rows, cols = len(mat), len(mat[0])
        res = []

        curr_col = curr_row = 0
        going_up = True

        while len(res) != rows * cols:
            if going_up:
                while curr_row >= 0 and curr_col < cols:
                    res.append(mat[curr_row][curr_col])

                    curr_row -= 1
                    curr_col += 1

                if curr_col == cols:
                    curr_col -= 1
                    curr_row += 2
                else:
                    curr_row += 1

                going_up = False
            else:
                while curr_row < rows and curr_col >= 0:
                    res.append(mat[curr_row][curr_col])

                    curr_col -= 1
                    curr_row += 1

                if curr_row == rows:
                    curr_col += 2
                    curr_row -= 1
                else:
                    curr_col += 1

                going_up = True

        return res